### ISSUE TITLE
Double pasting

### ISSUE THREAD

**Issue by ffMathy** [ffMathy/Shapeshifter#359]
The latest pre-release has an annoying double-paste bug. Will have to analyze logs closely.

**Comment by KoalaBear84** [comment]
Your ticket came just in time. Already was creating one. Good that it is reproducible.

There's a sequence of log lines that is logging it twice, from one CTRL + V operation. It looks like there is 2 interceptors picking it up and doing all of this twice.

**Comment by KoalaBear84** [comment]
Oops

**Comment by ffMathy** [comment]
Ooooh, so there's log data that can confirm this? Very interesting!

Can you post the relevant log sections?

**Comment by KoalaBear84** [comment]
This log is from a SINGLE CTRL + V, you can see it is processing it twice somehow.

[DoublePasting.txt](https://github.com/ffMathy/Shapeshifter/files/1965968/DoublePasting.txt)

**Comment by ffMathy** [comment]
There are two `WM_CLIPBOARDUPDATE` messages yes, but that is also expected. That is not the pasting itself, but the registration of new items in the clipboard.

One for when you originally copied, one for when Shapeshifter re-inserts the data to the clipboard.

There is only one actual paste:

> 2018-05-02 07:06:28.346 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardInjectionService] [INF]
> Clipboard package has been injected to the clipboard.
> 
> 2018-05-02 07:06:28.349 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardPasteService] [INF]
> Simulating paste with CTRL released and V released.
> 
> 2018-05-02 07:06:28.350 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardPasteService] [INF]
> Disabled paste hotkey interceptor.
> 
> 2018-05-02 07:06:28.351 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardPasteService] [INF]
> Uninstalled paste hotkey interceptor.
> 
> 2018-05-02 07:06:28.361 [#4448] [Shapeshifter.WindowsDesktop.Services.Keyboard.KeyboardManager] [INF]
> Sending key combinations [LeftCtrl: Down], [V: Down], [V: Up], [LeftCtrl: Up].
> 
> 2018-05-02 07:06:28.364 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardPasteService] [INF]
> Installed paste hotkey interceptor.
> 
> 2018-05-02 07:06:28.365 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardPasteService] [INF]
> Enabled paste hotkey interceptor.
> 
> 2018-05-02 07:06:28.366 [#4448] [Shapeshifter.WindowsDesktop.Services.Clipboard.ClipboardPasteService] [INF]
> Paste simulated.
> 
> 2018-05-02 07:06:28.366 [#4448] [Shapeshifter.WindowsDesktop.Data.Actions.PasteAction] [INF]
> Paste action performed.

**Comment by KoalaBear84** [comment]
Yes, but you do also see twice `Sending key combinations`.

**Comment by ffMathy** [comment]
I see! Well that should be very simple to resolve then! I don't have time until the weekend or so though.

Set a breakpoint where it emits that to the log, then right click the breakpoint and adjust its "hit count" to 2 (so that it will only break the 2nd time).

Then you should be able to compare the "Call stack" window of both the first and the second time, and be able to tell where the call is coming from, and why it is being duplicated.

Does that make sense?

**Comment by KoalaBear84** [comment]
Yes, makes sense. Will try to do that. Will probably be tomorrow because of #work 😀

**Comment by KoalaBear84** [comment]
Haha. Yes, do that! I know it is very hard, the hobby/work balance when you have a nice project / tool.

**Comment by ffMathy** [comment]
Can't wait to get time to fix this! It's the only bug that is currently preventing Shapeshifter from being surprisingly stable!
